### PR TITLE
Support expected errors with side effects

### PR DESCRIPTION
**Pull request by loveleif**
Enables asserting on side effects in failing queries. This is useful for `CALL ... IN TRANSACTIONS` queries, that will fail but also have side effects.

### CODE REVIEW

**Review by hvub — CHANGES_REQUESTED** [review]
The changes make sense in principle. However, from the scenario syntax point of view it is not obvious why additional syntax is needed. This seems to be only done because the steps are processed one-by-one `CypherTCK.toScenario`. 

I would rather prefer that a scenario does not have to ", with side effects:", because "Then an Error should be ... And the side effects should be. ..." already says this sufficiently. 

In `CypherTCK.toScenario` such scenarios can be detected by a one step look ahead. The look ahead can be implement, for example, by replacing 

```scala
val steps = pickle.getSteps.asScala.flatMap { step =>
```

with

```scala
    val stepsAsScala = pickle.getSteps.asScala.toSeq
    val successorSteps = stepsAsScala.tail.map(Some(_)) :+ None
    val stepsAsScalaWithSuccessor = stepsAsScala.zip(successorSteps)

    val steps = stepsAsScalaWithSuccessor.flatMap { case (step, successorStep) =>
```

then `successorStep` can be used to decide in `case expectErrorR(errorType, time, detail) => ...` whether to generate a `SideEffects(source = step).fillInZeros` or not.

I think, that given us a cleaner scenario syntax.

**Comment by loveleif** [comment]
> I would rather prefer that a scenario does not have to ", with side effects:", because "Then an Error should be ... And the side effects should be. ..." already says this sufficiently.

@hvub smart! I also didn't like it but didn't thought about that possibility, thanks!

**Comment by loveleif** [comment]
Forced pushed an improved solution to this branch.

**Review by hvub — APPROVED** [review]

**Comment by hvub** [comment]
I have looked at this again — commit to do the requested change myself — and found that it is not needed. 

Since `insertSideEffectsOnExpectError` sets the `source` of the inserted `SideEffects` steps to the `source` of the preceding `ExpectError` step and `ExpectError` steps always come with `StepType.THEN`, the inserted `SideEffects` steps fit the check in: 
https://github.com/opencypher/openCypher/blob/master/tools/tck-integrity-tests/src/test/scala/org/opencypher/tools/tck/ValidateSteps.scala#L110-L113


Bottom line: This PR can be merged.
